radar/exporter/exporters.py

```diff
@@ -357,96 +357,164 @@ def setup(self):
+                            self.ins_data['ins_biopsy_dia_label'] = ''
+                            row[31] = self.ins_data['ins_biopsy_dia_comments']
+                            self.ins_data['ins_biopsy_dia_comments'] = ''                   
+                            yield row
+                            row = self.makeRow(result, diagnosis)
+                            ins_data = self.getInsData(result, diagnosis)
+                        # New patient, no data found for previous patient
+                        else:
+                            row = self.makeRow(result, diagnosis)
+                            ins_data = self.getInsData(result, diagnosis)
+                
+                # Not new patient
+                else:
+                    # Primary diagnosis has been found, continue to check for INS data
+                    if row:
+                        ins_data = self.getInsData(result, diagnosis)
+                    # Search for primary diagnosis and INS data
+                    else:
+                        row = self.makeRow(result, diagnosis)
+                        ins_data = self.getInsData(result, diagnosis)
+        
+        # Not Nurture-ins export
+        else:            
+            for result in self._query:
+                diagnosis = result.diagnosis
+                if not diagnosis:
+                    continue
+                row = self.makeRow(result, diagnosis)
+                if row:
+                    yield row  
+
+    def makeRow(self, result, diagnosis):        
         
         if diagnosis not in self._primary:
             return
-        
-        self.row = [col[1](result) for col in self._columns]
-        
-        if diagnosis and diagnosis.codes:
-            for code in diagnosis.codes:
-                if code.system == "ERA-EDTA PRD":
-                    self.row[5] = code.code
-                if code.system == "ICD-10":
-                    self.row[6] = code.code
-                if code.system == "SNOMED CT":
-                    self.row[7] = code.code
+        else:
+            row = [col[1](result) for col in self._columns]
+            
+            if diagnosis and diagnosis.codes:
+                for code in diagnosis.codes:
+                    if code.system == "ERA-EDTA PRD":
+                        row[5] = code.code                        
+                    if code.system == "ICD-10":
+                        row[6] = code.code                        
+                    if code.system == "SNOMED CT":
+                        row[7] = code.code
+
+            return row
+
+    def getInsData(self, result, diagnosis):
+
+        for group_diagnosis in diagnosis.group_diagnoses:
+            if group_diagnosis.group.code == 'INS' and group_diagnosis.type.value == 'PRIMARY':
+                if not self.ins_data['ins_dia_date'] or self.ins_data['ins_dia_date'] < result.from_date:
```

Comment:
I don't think this is quite right. You're only populating self.ins_data['ins_dia_date'] from result.from_date so once it's not None it will never be less than result.from_date.

Reply:
It is possible for a patient to have more than one primary INS diagnosis so this just ensures it gets the most recent. IF there is no INS date or IF the current INS date is smaller than the result (result of the query) date then update the INS data otherwise leave it as it is. After each new patient, the INS data is wiped clean. Each patient can have many results.

Reply:
With the code as it is if diagnosis.group_diagnoses has multiple INS/PRIMARY entries then it will pick the first it comes to and set self.ins_data['ins_dia_date'] to result.from_date.

The check on self.ins_data['ins_dia_date'] < result.from_date will never trigger because it's always the same result.from_date value and so can never be less than itself.

Reply:
Doing a check / setting the value based on group_diagnosis.from_time would make sense if that's such a thing.

Reply:
I'm probably doing a bad job of explaining this. I only say that because I have tested this and it works. I do not think that from_time is a thing but I do think that from_date includes a timestamp. I would need to check to be 100%.

If I had 11 results 10 of which belong to the first patient and contain 2 INS diagnosis and the 11th belongs to another patient I believe it will follow this logic. Record 2 is INS diagnosis so data is added, the date is 01/01/2020. The 8th record is also INS data with date 02/01/2020. The logic gate would register false because there is already a from_date present from result 2. It then moves on to the second clause, compares and finds that result.from_date is larger than self.from_date and so enters and updates the data. If it was 01/12/2019 it would register false and exit. It then looks at record 11 which is for a different patient which sends it down the yield route.

All INS data is held through multiple loops until it finds a new patient number.

Reply:
I think I see now. It's a little odd having a condition within a loop that doesn't relate to the loop itself.

Assuming it's equivilent doing -

```if not self.ins_data['ins_dia_date'] or self.ins_data['ins_dia_date'] < result.from_date:
        self.getInsData(result, diagnosis)```

Would be clearer.

Reply:
Technically the function getInsData is not a loop but is called from within a loop, and contains the condition that is relevant to that function. I could move the condition outside of the function in the manner you suggest but the function is called in four separate places within a loop which I figured would mean repeating the condition unnecessarily. I thought it was best to avoid repeating code wherever possible.

Reply:
The most important thing is for the code to be as easy to read as possible, though code not being duplicated is often a consequence of that. It's also important, ideally, for functions to lend themselves to unit testing.

Without looking into the code - and in the absence of comments to the contrary - I would likely expect a function called "getInsData" to be just retrieving data and not applying conditions.

However, in the interests of getting this finished changing the function to -
```
def getInsData(self, result, diagnosis):
     '''Only get INS Data if no existing data or this diagnosis (result) is dated earlier than the earliest already seen'''
    if not self.ins_data['ins_dia_date'] or self.ins_data['ins_dia_date'] < result.from_date:
         for group_diagnosis in diagnosis.group_diagnoses:
```
Would still be an improvement and avoid confusion about it being in the loop.

Reply:
I will go and fix it as you suggest, I'd rather spend the time getting it right. Thanks for the feed back.